uat/testing-features/src/main/java/com/aws/greengrass/steps/CloudWatchSteps.java

```diff
@@ -80,34 +100,107 @@ public void deleteLogGroup(String componentType, String componentName) {
+                    logStreamName, logGroupName));
+        }
+    }
+
+
+    @Then("I verify {int} logs for {word} of type {word} have been uploaded to Cloudwatch within {int} seconds")
+    public void verifyLogs(int numberOfLogLines, String componentName, String componentType, int timeout) throws
+            Exception {
+        boolean result = waitSteps.untilTrue(() -> haveAllLogsBeenUploaded(numberOfLogLines, componentName,
+                componentType), timeout, TimeUnit.SECONDS);
+
+        if (!result) {
+            // Print all the cloudwatch logs it fetched, so we can debug what failed to get uploaded.
+            this.lastReceivedCloudWatchEvents.forEach(e -> LOGGER.info(e.message()));
+
+            throw new Exception(String.format("Failed to verify that %s logs were uploaded to CloudWatch",
+                    numberOfLogLines));
+        }
+    }
+
+
+    private boolean haveAllLogsBeenUploaded(int numberOfLogLines, String componentName, String componentType) {
+        // Logs written by the log generator append a sequence number per log line along with the component name
+        GetLogEventsRequest request = GetLogEventsRequest.builder()
+                .logGroupName(getLogGroupName(componentType, componentName))
+                .logStreamName(getLogStreamName())
+                .limit(numberOfLogLines) // limit of 10000 logs (this method could be optimized
+                .build();
+
+        LOGGER.info("Verifying {} logs present on group {} stream {}", numberOfLogLines, request.logGroupName(),
+                request.logStreamName());
+
+        try {
+            // The OTF watch steps check evey 100ms this to avoids hammering the api. Ideally OTF
+            // can allow us to configure the check interval rate
+            TimeUnit.SECONDS.sleep(5L);
+            GetLogEventsResponse response = cwClient.getLogEvents(request);
+            List<OutputLogEvent> events = response.events();
+
+            if (events.size() != numberOfLogLines) {
+                this.lastReceivedCloudWatchEvents = events;
+                return false;
+            }
+
+            // events is an unmodifiable list
+            List<OutputLogEvent> copy = new ArrayList<>(events);
+            copy.sort(Comparator.comparingLong(OutputLogEvent::timestamp));
+            return wereThereDuplicatesOrMisses(numberOfLogLines, componentName, copy);
+        } catch (ServiceUnavailableException | InterruptedException e) {
+            return false;
+        }
+    }
+
+    private boolean wereThereDuplicatesOrMisses(int expectedLogLines, String componentName,
+                                                List<OutputLogEvent> events) {
+
+        // Logs written by the log generator append a sequence number per log line along with the component name
+        for (int i = 1; i <= expectedLogLines; i++) {
+            String expected = String.format("(seq: %s, name: %s)", i, componentName);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: i (declared at line 188). This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 189. Consider formatting this data with &#39;%d&#39; instead.